lib/events.js

```diff
@@ -76,6 +76,74 @@ const kMaxEventTargetListeners = Symbol('events.maxEventTargetListeners');
 const kMaxEventTargetListenersWarned =
   Symbol('events.maxEventTargetListenersWarned');
 
+let EventEmitterAsyncResource;
+// The EventEmitterAsyncResource has to be initialized lazily because event.js
+// is loaded so early in the bootstrap process, before async_hooks is available.
+//
+// This implementation was adapted straight from addaleax's
+// eventemitter-asyncresource MIT-licensed userland module.
+// https://github.com/addaleax/eventemitter-asyncresource
+function lazyEventEmitterAsyncResource() {
+  if (EventEmitterAsyncResource === undefined) {
+    const {
+      AsyncResource
+    } = require('async_hooks');
+
+    const kEventEmitter = Symbol('kEventEmitter');
+    const kAsyncResource = Symbol('kAsyncResource');
+    class EventEmitterReferencingAsyncResource extends AsyncResource {
+      constructor(ee, type, options) {
+        super(type, options);
+        this[kEventEmitter] = ee;
+      }
+
+      get eventEmitter() {
+        return this[kEventEmitter];
+      }
+    }
+
+    EventEmitterAsyncResource =
+      class EventEmitterAsyncResource extends EventEmitter {
+        constructor(options = undefined) {
+          let name;
+          if (typeof options === 'string') {
+            name = options;
+            options = undefined;
+          } else {
+            name = options?.name || new.target.name;
```

Comment:
I think name should be mandatory instead of defaulting to `new.target.name` similar as it is mandatory in `AsyncResource`.

Reply:
I disagree. I think there would be no problem with defaulting the type in `AsyncResource` also. /cc @addaleax since she's the one who made this choice originally.

Reply:
If it would be mandatory to create a derived class I would agree. But both classes can be used directly which results in quite a risk that independent resource with the same name are created just because people forget to specify a name (or have a typo like `options.nname`.

Reply:
No strong feelings here :) We could also only require an explicit name if `new.target === EventEmitterAsyncResource` :woman_shrugging:

Reply:
I think that's a good compromise

Reply:
Updated with @addaleax's suggestion